### PR TITLE
Add support for CodeQL

### LINKED ISSUE #2655: Analyze add-on source code to detect vulnerabilities
### Feature request

To improve add-ons security, the source code should be analized using automated tools to detect vulnerabilities, as @adriani90 suggested in this [discussion about security for add-ons](https://github.com/nvaccess/nvda/discussions/16241).

### Suggestions

Existing GitHub Actions to use CodeQl maybe used, unzipping the add-on source code after creating the pull request when an add-on is submitted.
This requires that add-ons include the source code in the binary file.
The CodeQl action shouldn't be configured just for Python, since add-ons may include other languages, like Rust, used by @LeonarddeR in RD Access, or readFeeds, used by me in readFeeds, just to point out some examples.

### PR DESCRIPTION
**Pull request by nvdaes**
### Issue number

fixes #2655

### Summary of the issue

Some vulnerabilities can be detected analyzing source code of add-ons submitted to the store.

### Development strategy

1. After validating add-on metadata, Upload artifact action is used to upload the add-on previously downloaded to be validated.
1.  A workflow to analyze the source code with CodeQL is called.
1. This workflow calls the Download artifact action to download the add-on.
1. The addon.nvda-addon file is moved to an addon.zip file to be expanded for the analysis.
1. The source code is analyzed. Languages supported in the used workflow are just Python and JavaScript, since they are interpreted languages which can be used in add-ons. Other languages like C++ may require .dll files, so maybe faster just to analyze the mentioned languages.
1. If analysis fails, a comment is posted to the issue requesting the submitter to contact NV Access for more details (failures aren't tested now).
1. If the job succeeds, the workflow to check metadata and create a pull request with the .json file is called.

### CODE REVIEW

**Comment by nvdaes** [comment]
@seanbudd, This has been tested in

https://github.com/nvdaes/addon-datastore/actions/runs/8287950825

**Comment by seanbudd** [comment]
@nvdaes - can you show an example of a failure?

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      # ℹ️ Command-line programs to run using the OS shell.
+      # 📚 https://git.io/JvXDl
+
+      # ✏️ If the Autobuild fails above, remove it and uncomment the following three lines
+      #    and modify them (or add more) to build your code if your project
+      #    uses a compiled language
+
+      #- run: |
+      #   make bootstrap
+      #   make release
```
Comment (seanbudd) [inline .github/workflows/codeql-analysis.yml:71]: are these comments relevant to our repository?

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+      - name: Perform CodeQL Analysis
+        uses: github/codeql-action/analyze@v3
+      - name: Warn if analysis fails
+        if: failure()
+        uses: peter-evans/create-or-update-comment@v4
+        with:
+          issue-number: ${{ github.event.issue.number }}
+          body: |
+            CodeQL analysis has failed for this add-on.
```
Comment (seanbudd) [inline .github/workflows/codeql-analysis.yml:81]:
```suggestion
            Security analysis has failed for this add-on.
```

File: .github/workflows/codeql-analysis.yml [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#
+# You may wish to alter this file to override the set of languages analyzed,
+# or to provide custom queries or build logic.
+#
+# ******** NOTE ********
+# We have attempted to detect the languages in your repository. Please check
+# the `language` matrix defined below to confirm you have the correct set of
+# supported CodeQL languages.
+#
+name: CodeQL
```
Comment (seanbudd) [inline .github/workflows/codeql-analysis.yml:12]:
```suggestion
name: CodeQL security analysis
```

File: .github/workflows/codeql-analysis.yml [hunk 4]
```diff
@@ -0,0 +1,91 @@
+# For most projects, this workflow file will not need changing; you simply need
+# to commit it to your repository.
+#
+# You may wish to alter this file to override the set of languages analyzed,
+# or to provide custom queries or build logic.
```
Comment (seanbudd) [inline .github/workflows/codeql-analysis.yml:5]:
```suggestion
# Note, the base of this comes from an autogenerated file from CodeQL
```

File: .github/workflows/codeql-analysis.yml (hunk 1 above)
Comment (nvdaes) [reply .github/workflows/codeql-analysis.yml:67]: No, I'm remove them. This code is based on the Javascript action template used for my build-discussion action, currently used in the store repo, and I supposed that comments would make easier to review changes adding some info about CodeQL.

**Comment by nvdaes** [comment]
> @nvdaes - can you show an example of a failure?

Unfortunately, the CodeQL action seems to fail just when is triggered in a PR, not for other events for example not when the code is analized on workflow_call or schedule events.
Anyway, if it detects a failure, an alert can be found in the security page of the repo, so that NV Access may check these alerts for add-ons submitted to the store and request authors to fix them or remove the add-on.

**Comment by nvdaes** [comment]
We may use the .sarif files generated by the analyze action to see if the results property contains elements, and then post a comment in the corresponding issue requesting authors to contact NV Access. I'll convert this to a draft for now.

**Comment by nvdaes** [comment]
I think this can be reviewed again.
Since analysis was always successful,even when errors were found, since it seems to fail just for pull requests or whendifferences in code ara analized,not when analysis is performed for a set of files for example on workflow_call, I've created a securityAnalysis.js file to determine if erors are present in results (with default configuration, since a deeper analysis may produce more false positive results).
Iferrors are found, the jobwill fail now and an issue commentwill be post, and the issue will be closed. In this case, an artifact with a sarif file will be uploaded to the action, since I don't know if reading lots of alerts in the security scanning page of GitHub will be very comfortable or accessible. In this way authors may see the produced artifact or NV Access may provide it.
We may also provide the artifact URL in the issue.
I don't think that this is violating security reporting good practices, since codeQL is publicly available and authors should know about this, that maybedocumented in the readme or in the issue template.
Also, if authors don't include source code in their add-ons, security analysis will be successfull. We may check the sarif files to make a warning somewere if no files are analyzed for Python and Javascript.

I've tested this in the following workflow runs:

- [Add-on containing Python source code without failures related to code analysis](https://github.com/nvdaes/addon-datastore/actions/runs/8310707532)
- [Add-on without errors, just with cpython files](https://github.com/nvdaes/addon-datastore/actions/runs/8310659141)
- [Add-on with error in Python file, contains good JavaScript](https://github.com/nvdaes/addon-datastore/actions/runs/8310408328)
- [Add-on without errors](https://github.com/nvdaes/addon-datastore/actions/runs/8310532054)

Note that some workflows fail due to things not related to analysis.

**Comment by nvdaes** [comment]
cc: @CyrilleB79 (security reporter and analyst)

**Comment by XLTechie** [comment]
> Note that some workflows fail due to things not related to analysis.

@nvdaes Given that you propose blocking submission on failure, can you please 
explain this point in more detail?

**Comment by nvdaes** [comment]
> @nvdaes Given that you propose blocking submission on failure, can you please
explain this point in more detail?

I was meaning workflows on my fork of the store, shown here as examples of tests made by me. Some workflows may fail if the used add-on is already on my fork of the store, and I mentioned it to indicate that the important thing in those cases is to pay attention to the workflow until the analysis job.
Hope this clarifies what I wanted to mean. This was related to my tests, not to the pull request itself.

**Comment by seanbudd** [comment]
Can you please add documentation on how to handle these failures as a reviewer?
For example, I'd encourage using a Sarif Viewer, particularly the VS Code one:  https://marketplace.visualstudio.com/items?itemName=MS-SarifVSCode.sarif-viewer

Even better, can these Sarif files become viewable in GitHub?
From what I can understand, these can be uploaded via [these steps](https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning) and viewed at a URL like `https://github.com/nvaccess/nvda/security/code-scanning`

File: .github/workflows/codeql-analysis.yml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+# supported CodeQL languages.
+#
+name: CodeQL security analysis
+
+on:
+  workflow_call:
+
+jobs:
+  analyze:
+    name: Analyze
```
Comment (seanbudd) [inline .github/workflows/codeql-analysis.yml:15]:
```suggestion
    name: Analyze add-on
```

**Review by seanbudd — APPROVED** [review]
Code changes look good, just needs documentation

**Comment by CyrilleB79** [comment]
Do you confirm that this check is mandatory to have an add-on included or updated in the store? If yes, this probably needs to be documented. Thanks.

**Comment by seanbudd** [comment]
I would say that any failures will need to be manually overruled by NV Access

**Comment by nvdaes** [comment]
Sean wrote:

> I would say that any failures will need to be manually overruled by NV Access

I think this is a good idea. Anyway I'll try to add documentation in the submitters guide. I think this is the best place, and, after testing the VS Code extension, I'll add a note about it, as well as about how to see sarifs on GitHub, though I don't know if this would be fully accessible, deppending of the number of errors and so on.

**Comment by XLTechie** [comment]
Is there anything other than VS Code extension (which many/most add-on devs 
don't use), or "possibly inaccessible" GitHub solution?

**Comment by nvdaes** [comment]
> Is there anything other than VS Code extension (which many/most add-on devs
don't use), or "possibly inaccessible" GitHub solution?

According to my tests, in the security code scanning page of GitHub, alerts are shown like issues and we can press enter in the alert title.
Example:

Title: Incomplete URL substring sanitization 

High 
#1 opened 
Body of the alert:

Incomplete URL substring sanitization 
High 
#1 opened last month • Detected by CodeQL in addon/.../readFeeds/__init__.py:681 
master
 
Also, since this maybe incomplete, when the analysis fails, in this PR we make that GitHub Actions download the add-on and the sarif files containing failures to the artifact sections of the GitHub workflow.
Testing VS Code, I feel that the Sarif Viewer extension can be used, but this is not friendly. I moved to the status bar, pressed a related button and so on.
I prefer to open the sarif file with Notepad++, or with VS Code, and search with control+f, for example, for startline, to see where failures are found. I'll attach here an artifact from readFeeds 26.1.0.

**Comment by nvdaes** [comment]
[results-python.zip](https://github.com/nvaccess/addon-datastore/files/14977074/results-python.zip)
This is a sample artifact showing an error
[results-python.zip](https://github.com/nvaccess/addon-datastore/files/14977093/results-python.zip)

**Comment by nvdaes** [comment]
I'll mention this VS Code extension.
Whe a sarif file is opened, we can press NVDA+space and then t, so that NVDA moves to the table of errors. There, pressing enter in a clickable result, more details have been shown.
Extension:

Name: SARIF Explorer
Id: trailofbits.sarif-explorer
Description: SARIF Explorer: Explore static analysis results effectively and enjoyably.
Version: 1.1.0
Publisher: Trail of Bits
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=trailofbits.sarif-explorer

Output after clicking:

py/incomplete-url-substring-sanitization
1
 
 
Result marked as TODO
__init__.py:681
The string https://www.google.com may be at an arbitrary position in the sanitized URL.
 
No result selected

**Comment by nvdaes** [comment]
I've made a commit to add the GitHub workflow link to the issue when the bot comments about code analysis failure, so that documenting and seing results and artifacts can be easier.
Otherwise, in the documentation we will need to say something about "the corresponding workflow", and this would be confusing.

See tests for this at

https://github.com/nvdaes/addon-datastore/issues/999

**Comment by nvdaes** [comment]
I think we should remove the info about Sarif Explorer extension for VS Code. I don't get consistent results across different computers, and now I cannot use the extension after installing it.
For me the best approach is to open the sarif file in Notepad++, or in VS Code, searching text like the word "region" to see where the errors are located.

File: README.md [hunk 6]
```diff
@@ -15,8 +15,13 @@ For an overview of the whole Add-on Store, read [the design overview](./docs/des
 ### About security 
 Ensuring that an add-on is safe to run is a difficult challenge that isn't addressed here.
 However, the metadata for a new submission (add-on release) can be confirmed to match its manifest description.
```
Comment (seanbudd) [inline README.md:17]: thoughts on removing these 2 sentences and replacing it with something like?
It seems like we are trying to tackle the first sentence now, and the second sentence doesn't really seem relevant to security.

>Add-ons are run at user's own risk, add-ons in the add-on store do not undergo human security audits.
>The add-on store includes the following security measures:

File: docs/submitters/submissionGuide.md [hunk 7]
```diff
@@ -8,6 +8,9 @@ Submitting an add-on version is done via a GitHub issue form.
 A JSON metadata file is generated from the issue form and the add-on's manifest.
 A detailed description of the JSON schema can be found in [jsonMetadata.md](./jsonMetadata.md).
 When the issue form is submitted, the JSON file is generated and submitted as a pull request to the repository.
+Security analysis is performedfor the submitted add-on.
```
Comment (seanbudd) [inline docs/submitters/submissionGuide.md:11]:
```suggestion
Security analysis is performed for the submitted add-on.
```

File: docs/submitters/submissionGuide.md [hunk 8]
```diff
@@ -42,3 +45,4 @@ The add-on should soon become available in the store.
 Optional.
 Some authors may wish to include their add-on in the translations system.
 To do this, follow the steps to register the add-on in [mrconfig](https://github.com/nvaccess/mrconfig/blob/master/readme.md#steps-for-addon-authors).
+for
```
Comment (seanbudd) [inline docs/submitters/submissionGuide.md:48]:
```suggestion
```

File: docs/submitters/submissionGuide.md [hunk 9]
```diff
@@ -8,6 +8,9 @@ Submitting an add-on version is done via a GitHub issue form.
 A JSON metadata file is generated from the issue form and the add-on's manifest.
 A detailed description of the JSON schema can be found in [jsonMetadata.md](./jsonMetadata.md).
 When the issue form is submitted, the JSON file is generated and submitted as a pull request to the repository.
+Security analysis is performedfor the submitted add-on.
+If there are detected vulnerabilities, a link to a GitHub workflow will be shown in the issue. You can open this link and download artifacts containing the results of the analysis.
+Unzip artifacts and open the .sarif file in your preferred editor. Then, you can search properties like "region" to investigate where vulnerabilities are located in your code.
```
Comment (seanbudd) [inline docs/submitters/submissionGuide.md:13]: I really think reading the RAW sarif format is challenging for users.
I have suggested an alternative, a web based sarif viewer from microsoft. It appears accessible with NVDA+Chrome.

```suggestion
If there are detected vulnerabilities, a link to a GitHub workflow will be shown in the issue.
You can open this link and download artifacts containing the results of the analysis.
Unzip artifacts and open the .sarif file in your preferred editor.
For example, you can use [Microsoft's Sarif web based reader](https://microsoft.github.io/sarif-web-component/).
```

**Review by seanbudd — CHANGES_REQUESTED** [review]

**Comment by nvdaes** [comment]
I'll apply all suggested changes, including the quotation about security.
I've tested the sarif viewer with Firefox and I'll use it for myself, so I think it's a good idea.

File: docs/submitters/submissionGuide.md [hunk 10]
```diff
@@ -8,9 +8,11 @@ Submitting an add-on version is done via a GitHub issue form.
...
 A detailed description of the JSON schema can be found in [jsonMetadata.md](./jsonMetadata.md).
 When the issue form is submitted, the JSON file is generated and submitted as a pull request to the repository.
-Security analysis is performedfor the submitted add-on.
-If there are detected vulnerabilities, a link to a GitHub workflow will be shown in the issue. You can open this link and download artifacts containing the results of the analysis.
-Unzip artifacts and open the .sarif file in your preferred editor. Then, you can search properties like "region" to investigate where vulnerabilities are located in your code.
+Security analysis is performed for the submitted add-on.
+If there are detected vulnerabilities, a link to a GitHub workflow will be shown in the issue.
+You can open this link and download artifacts containing the results of the analysis.
+Unzip artifacts and open the .sarif file in your preferred editor.
+For example, you can use [Microsoft's Sarif web based reader](https://microsoft.github.io/sarif-web-component/).
```
Comment (seanbudd) [inline docs/submitters/submissionGuide.md:15]:
```suggestion
For example, you can use [Microsoft's Sarif web based reader](https://microsoft.github.io/sarif-web-component/).
NV Access will determine whether or not the detection should prevent the add-on from being accepted.
Please review the warnings and consider whether you want to fix this in the add-on.
If you can provide more context on the failure in the submission, please do.
```

File: README.md (hunk 6 above)
Comment (seanbudd) [reply README.md:17]: any thoughts on this change?

**Comment by nvdaes** [comment]
@seanbudd, I think this is ready.

**Review by seanbudd — APPROVED** [review]
Thanks @nvdaes

**Comment by nvdaes** [comment]
Thanks for merging this, @seanbud.
I think we should inform about this on the NVDA add-ons mailing list, so that people is aware of this change and know that their code can be analized when submitting add-ons.
I'll do it tomorrow, if you don't prefer to do it on behalf of NV Access.
